distributed/scheduler.py

```diff
@@ -1576,3919 +1566,4196 @@ class Scheduler(ServerNode):
-            ("no-worker", "waiting"): self.transition_no_worker_waiting,
-            ("released", "forgotten"): self.transition_released_forgotten,
-            ("memory", "forgotten"): self.transition_memory_forgotten,
-            ("erred", "forgotten"): self.transition_released_forgotten,
-            ("erred", "released"): self.transition_erred_released,
-            ("memory", "released"): self.transition_memory_released,
-            ("released", "erred"): self.transition_released_erred,
-        }
+            ts.state = "waiting"
 
-        connection_limit = get_fileno_limit() / 2
+            recommendations: dict = {}
 
-        super().__init__(
-            handlers=self.handlers,
-            stream_handlers=merge(worker_handlers, client_handlers),
-            io_loop=self.loop,
-            connection_limit=connection_limit,
-            deserialize=False,
-            connection_args=self.connection_args,
-            **kwargs,
-        )
+            dts: TaskState
+            for dts in ts._dependencies:
+                if dts._exception_blame:
+                    ts._exception_blame = dts._exception_blame
+                    recommendations[key] = "erred"
+                    return recommendations, worker_msgs, client_msgs
 
-        if self.worker_ttl:
-            pc = PeriodicCallback(self.check_worker_ttl, self.worker_ttl)
-            self.periodic_callbacks["worker-ttl"] = pc
+            for dts in ts._dependencies:
+                dep = dts._key
+                if not dts._who_has:
+                    ts._waiting_on.add(dts)
+                if dts._state == "released":
+                    recommendations[dep] = "waiting"
+                else:
+                    dts._waiters.add(ts)
 
-        if self.idle_timeout:
-            pc = PeriodicCallback(self.check_idle, self.idle_timeout / 4)
-            self.periodic_callbacks["idle-timeout"] = pc
+            ts._waiters = {dts for dts in ts._dependents if dts._state == "waiting"}
 
-        if extensions is None:
-            extensions = list(DEFAULT_EXTENSIONS)
-            if dask.config.get("distributed.scheduler.work-stealing"):
-                extensions.append(WorkStealing)
-        for ext in extensions:
-            ext(self)
+            if not ts._waiting_on:
+                if self._workers_dv:
+                    recommendations[key] = "processing"
+                else:
+                    self._unrunnable.add(ts)
+                    ts.state = "no-worker"
 
-        setproctitle("dask-scheduler [not started]")
-        Scheduler._instances.add(self)
-        self.rpc.allow_offload = False
-        self.status = Status.undefined
+            return recommendations, worker_msgs, client_msgs
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
 
-    ##################
-    # Administration #
-    ##################
+                pdb.set_trace()
+            raise
```

Comment:
I suggest we implement all the exception logging using a context manager as described in https://stackoverflow.com/a/28158006

Reply:
Yeah I agree that something like that makes sense. Alternatively we might want to push this into `transition` where all of these methods are called. Held off on that for now as I didn't want to disrupt the code too much, but agree this is something we should follow up on.